### ISSUE TITLE
Calls to assert_that slow down integrate_to_ppi

### ISSUE THREAD

**Issue by bart1** [adokter/bioRad#358]
Repeatedly checking input parameter considerably slow down `integrate_to_ppi` in the example below it takes 30% of the total execution time. A quick check shows `project_to_ppi` does not have any issues

``` r
require(bioRad)
#> Loading required package: bioRad
#> Welcome to bioRad version 0.5.1.9344
#> Docker daemon running, Docker functionality enabled (vol2bird version 0.5.0)
sessionInfo()
#> R version 3.6.3 (2020-02-29)
#> Platform: x86_64-pc-linux-gnu (64-bit)
#> Running under: Ubuntu 18.04.4 LTS
#> 
#> Matrix products: default
#> BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
#> LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1
#> 
#> locale:
#>  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
#>  [3] LC_TIME=nl_NL.UTF-8        LC_COLLATE=en_US.UTF-8    
#>  [5] LC_MONETARY=nl_NL.UTF-8    LC_MESSAGES=en_US.UTF-8   
#>  [7] LC_PAPER=nl_NL.UTF-8       LC_NAME=C                 
#>  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
#> [11] LC_MEASUREMENT=nl_NL.UTF-8 LC_IDENTIFICATION=C       
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] bioRad_0.5.1.9344
#> 
#> loaded via a namespace (and not attached):
#>  [1] spam_2.5-1          tidyselect_1.0.0    xfun_0.13          
#>  [4] purrr_0.3.3         lattice_0.20-40     rhdf5_2.30.1       
#>  [7] colorspace_1.4-1    vctrs_0.2.4         htmltools_0.4.0    
#> [10] yaml_2.2.1          rlang_0.4.5         pillar_1.4.3       
#> [13] foreign_0.8-75      glue_1.4.0          sp_1.4-1           
#> [16] jpeg_0.1-8.1        lifecycle_0.2.0     plyr_1.8.6         
#> [19] stringr_1.4.0       fields_10.3         dotCall64_1.0-0    
#> [22] munsell_0.5.0       gtable_0.3.0        raster_3.0-12      
#> [25] RgoogleMaps_1.4.5.3 codetools_0.2-16    evaluate_0.14      
#> [28] knitr_1.28          maptools_0.9-9      curl_4.3           
#> [31] fansi_0.4.1         highr_0.8           Rcpp_1.0.4.6       
#> [34] scales_1.1.0        rjson_0.2.20        ggplot2_3.3.0      
#> [37] png_0.1-7           digest_0.6.25       stringi_1.4.6      
#> [40] dplyr_0.8.5         grid_3.6.3          rgdal_1.4-8        
#> [43] cli_2.0.2           tools_3.6.3         bitops_1.0-6       
#> [46] magrittr_1.5        maps_3.3.0          tibble_3.0.0       
#> [49] crayon_1.3.4.9000   tidyr_1.0.2         pkgconfig_2.0.3    
#> [52] ellipsis_0.3.0      data.table_1.12.8   assertthat_0.2.1   
#> [55] rmarkdown_2.1       httr_1.4.1          Rhdf5lib_1.8.0     
#> [58] R6_2.4.1            ggmap_3.0.0         compiler_3.6.3

pvolfile <- system.file("extdata", "volume.h5", package = "bioRad")

example_pvol <- read_pvolfile(pvolfile)

# load the corresponding vertical profile for this polar volume
data(example_vp)
Rprof(interval=.01)
my_ppi <- integrate_to_ppi(example_pvol, example_vp, nx = 50, ny = 50)
Rprof(NULL)
r<-summaryRprof()
r$by.total['"assert_that"',]
#>               total.time total.pct self.time self.pct
#> "assert_that"       0.72     32.58      0.03     1.36
# A large percentage of the total time is spend in an assert_that call
l<-readLines('Rprof.out')
# These calls seem to occur in 4 functions about equally frequent
table(unlist(lapply(strsplit(l, ' '), function(x) x[which(x=='"assert_that"')+1])))
#> 
#>           "beam_height"          "beam_profile"            "beam_width" 
#>                      23                      14                       3 
#> "gaussian_beam_profile" 
#>                      32
```

<sup>Created on 2020-04-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by peterdesmet** [comment]
This is `assert_that()` for the `integrate_to_ppi()` function or one of the called functions?

**Comment by bart1** [comment]
Making internal functions hardly slows down normal calls, but calls to an internal function are a lot faster (they take less then ten percent):
``` r
require(assertthat)
#> Loading required package: assertthat
beam_height_old <- function(range, elev, k = 4 / 3, lat = 35, re = 6378,
                            rp = 6357) {
  assert_that(is.numeric(range))
  assert_that(is.numeric(elev))
  assert_that(is.number(k))
  assert_that(is.number(lat))
  assert_that(is.number(re))
  assert_that(is.number(rp))
  sqrt(
    range^2 + (k * earth_radius(re, rp, lat))^2 +
      2 * range * (k * earth_radius(re, rp, lat)) * sin(elev * pi / 180)
  ) - k * earth_radius(re, rp, lat)}

beam_height_new <- function(range, elev, k = 4 / 3, lat = 35, re = 6378,
                        rp = 6357) {
  assert_that(is.numeric(range))
  assert_that(is.numeric(elev))
  assert_that(is.number(k))
  assert_that(is.number(lat))
  assert_that(is.number(re))
  assert_that(is.number(rp))
  beam_height_internal(range = range, elev = elev, k = k, lat = lat, re = re, rp = rp)
}

beam_height_internal <- function(range, elev, k = 4 / 3, lat = 35, re = 6378,
                                 rp = 6357) {
  sqrt(
    range^2 + (k * earth_radius(re, rp, lat))^2 +
      2 * range * (k * earth_radius(re, rp, lat)) * sin(elev * pi / 180)
  ) - k * earth_radius(re, rp, lat)
}

earth_radius <- function(a, b, lat) {
  lat <- lat * pi / 180
  1000 * sqrt(
    ((a^2 * cos(lat))^2 + (b^2 * sin(lat))^2) /
      ((a * cos(lat))^2 + (b * sin(lat))^2)
  )
}

microbenchmark::microbenchmark(
  beam_height_new(1:100, 2),
  beam_height_old(1:100, 2),
  beam_height_internal(1:100, 2)
)
#> Unit: microseconds
#>                            expr    min       lq     mean   median       uq
#>       beam_height_new(1:100, 2) 96.310 101.9755 302.7482 132.4190 135.4955
#>       beam_height_old(1:100, 2) 95.869 101.5445 496.3035 130.4385 134.8925
#>  beam_height_internal(1:100, 2)  5.956   6.8730   8.5557   7.5930   9.9385
#>        max neval cld
#>   8487.984   100   a
#>  36942.014   100   a
#>     16.338   100   a
```

<sup>Created on 2020-04-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by bart1** [comment]
@peterdesmet The calls occur in these functions: "beam_height"   "beam_profile"  "beam_width"  "gaussian_beam_profile" are the once that get caught in the profiling

**Comment by bart1** [comment]
this is now partially addressed in 5ef6ebca4aa78b232ef20fab9b6ba9db69e41dcb . There is now an internal function for `beam_height`, `beam_width` and `gaussian_beam_profile` as for these cases it is easy to confirm the checks have already been run in functions calling these functions. It still needs to be checked for beam_profile if all checks have already been run for calls in `beam_profile_overlap_help` /  `beam_profile_overlap`

**Comment by bart1** [comment]
@peterdesmet the later two function might also require some unit tests. Do you think their checking also should be changed to `assert_that`? If so it might be good to do that before changing to internal functions.
